src/nodes/modbusServer.js

```diff
@@ -39,20 +39,54 @@ module.exports = function (RED) {
         this.getDiscreteInput = function (register, receive, errorHandler) {
             node.addRoute(register, 'getDiscreteInput', receive, errorHandler);
         };
+        this.setRegister = function (register, receive, errorHandler) {
+            node.addRoute(register, 'setRegister', receive, errorHandler);
+        };
+        this.setCoil = function (register, receive, errorHandler) {
+            node.addRoute(register, 'setCoil', receive, errorHandler);
+        };
         const apiMethods = {
-            _apiHandler: function (addr, command, unitID, callback, defaultval) {
+            _apiReadHandler: function (addr, command, unitID, callback, defaultval) {
                 let handler = node.routes.get(`${command}.${addr}`);
-                let req = {
+                    let req = {
                     register: addr,
                     unitID: unitID,
                     command: command
                 };
                 let res = {
                     callback: callback,
                 };
-                if (handler) handler.receive(req, res);
-                else {
-                    return callback(null, defaultval);
+                if (handler) {
+                    handler.receive(req, res);
+                } else {
+                    callback({
+                        modbusErrorCode: 0x02, // Illegal address
+                        msg: "Invalid length"
+                    },null);
+                    //node.error({error:'A client attempted to read from invalid register',req:req});
+                    node.error('A client attempted to read from invalid register '+ JSON.stringify(req,null,'\t'));
+                }
+            },
+            _apiWriteHandler: function (addr, command, unitID, callback, value) {
+                let handler = node.routes.get(`${command}.${addr}`);
+                let req = {
+                    register: addr,
+                    unitID: unitID,
+                    command: command,
+                    value: value,
+                };
+                let res = {
+                    callback: callback,
+                };
+                if (handler){
+                    handler.receive(req, res);
+                } else {
+                    callback({
+                        modbusErrorCode: 0x02, // Illegal address
+                        msg: "Invalid length"
+                    },null);
+                    //node.error({error:'A client attempted to write to invalid register',req:req});
+                    node.error('A client attempted to write to invalid register '+ JSON.stringify(req,null,'\t'));
```

Comment:
Which of these lines would you prefer having in the final merge?

Reply:
- In node-red 1.0, the categories changed so that there are no long any "input" or "output" categories. http, websocket and similar nodes are now under "network". This left your nodes under two categories containing only one node, which is why I changed it. Forgive me for including that change in this pull request. Maybe a better option is to categorize both under "modbus"?
- As for testing; I can provide the flow I used to test in node-red, obviously would have this node version as a dependency. I also used node-red-contrib-moment and node-red-contrib-modbus. I did not write any test outside of the node-red flow editor. Here's the flow I used:
`[{"id":"59303127.37f18","type":"tab","label":"Flow 1","disabled":false,"info":""},{"id":"a35390cd.0fbb4","type":"modbus in","z":"59303127.37f18","name":"","modbusServer":"94b02819.8ff308","command":"getInputRegister","register":"100","x":330,"y":300,"wires":[["d28d0b68.edb768"]]},{"id":"9741c2a3.f204a","type":"modbus out","z":"59303127.37f18","name":"","x":1090,"y":600,"wires":[]},{"id":"c2542f68.6c5fb","type":"modbus in","z":"59303127.37f18","name":"","modbusServer":"94b02819.8ff308","command":"getInputRegister","register":"101","x":330,"y":360,"wires":[["6fcb9fdd.13cd"]]},{"id":"74d40767.4a4f98","type":"modbus in","z":"59303127.37f18","name":"","modbusServer":"94b02819.8ff308","command":"getInputRegister","register":"102","x":330,"y":420,"wires":[["ce94b021.ab2a2"]]},{"id":"f301d304.216bf","type":"modbus in","z":"59303127.37f18","name":"","modbusServer":"94b02819.8ff308","command":"getHoldingRegister","register":"103","x":340,"y":480,"wires":[["78a1ae1.bdaf35"]]},{"id":"9aa44506.7e6598","type":"modbustcp-read","z":"59303127.37f18","name":"","topic":"","dataType":"InputRegister","adr":"100","quantity":"3","rate":"1","rateUnit":"s","server":"ecc1bc26.f60bd","ieeeType":"off","ieeeBE":"true","x":280,"y":100,"wires":[["82846267.02a39"]]},{"id":"82846267.02a39","type":"debug","z":"59303127.37f18","name":"","active":false,"tosidebar":true,"console":false,"tostatus":false,"complete":"false","x":510,"y":100,"wires":[]},{"id":"1aaa5da2.d79182","type":"modbustcp-write","z":"59303127.37f18","name":"","topic":"","dataType":"HoldingRegister","adr":"104","server":"ecc1bc26.f60bd","x":710,"y":180,"wires":[]},{"id":"cf7b273a.ed9b78","type":"inject","z":"59303127.37f18","name":"","topic":"","payload":"3","payloadType":"num","repeat":"1","crontab":"","once":false,"onceDelay":0.1,"x":270,"y":180,"wires":[["3bc2e903.9b6b86"]]},{"id":"d6c7244e.ed8e48","type":"modbus in","z":"59303127.37f18","name":"","modbusServer":"94b02819.8ff308","command":"getHoldingRegister","register":"104","x":340,"y":540,"wires":[["cba2eb8e.3414f8"]]},{"id":"e2c9ec96.8d0b2","type":"modbus in","z":"59303127.37f18","name":"","modbusServer":"94b02819.8ff308","command":"setRegister","register":"104","x":320,"y":680,"wires":[["40c9976f.5f1c08"]]},{"id":"cba2eb8e.3414f8","type":"change","z":"59303127.37f18","name":"","rules":[{"t":"set","p":"payload","pt":"msg","to":"register104","tot":"global"}],"action":"","property":"","from":"","to":"","reg":false,"x":580,"y":540,"wires":[["9741c2a3.f204a"]]},{"id":"40c9976f.5f1c08","type":"change","z":"59303127.37f18","name":"","rules":[{"t":"set","p":"register104","pt":"global","to":"payload","tot":"msg"}],"action":"","property":"","from":"","to":"","reg":false,"x":580,"y":680,"wires":[["9741c2a3.f204a"]]},{"id":"55240f4b.b5c24","type":"moment","z":"59303127.37f18","name":"","topic":"","input":"payload","inputType":"msg","inTz":"America/Los_Angeles","adjAmount":0,"adjType":"days","adjDir":"add","format":"Hmm","locale":"en_US","output":"payload","outputType":"msg","outTz":"America/Los_Angeles","x":760,"y":300,"wires":[["9741c2a3.f204a"]]},{"id":"d28d0b68.edb768","type":"change","z":"59303127.37f18","name":"","rules":[{"t":"set","p":"payload","pt":"msg","to":"","tot":"date"}],"action":"","property":"","from":"","to":"","reg":false,"x":560,"y":300,"wires":[["55240f4b.b5c24"]]},{"id":"64f10c27.2d0f84","type":"modbus in","z":"59303127.37f18","name":"","modbusServer":"94b02819.8ff308","command":"setRegister","register":"103","x":320,"y":600,"wires":[["853c2955.9cc818"]]},{"id":"853c2955.9cc818","type":"change","z":"59303127.37f18","name":"","rules":[{"t":"set","p":"register103","pt":"global","to":"payload","tot":"msg"}],"action":"","property":"","from":"","to":"","reg":false,"x":580,"y":600,"wires":[["9741c2a3.f204a"]]},{"id":"6fcb9fdd.13cd","type":"random","z":"59303127.37f18","name":"","low":"1","high":"16000","inte":"true","property":"payload","x":540,"y":360,"wires":[["38fd460e.21533a"]]},{"id":"38fd460e.21533a","type":"change","z":"59303127.37f18","name":"","rules":[{"t":"set","p":"register102","pt":"global","to":"payload","tot":"msg"}],"action":"","property":"","from":"","to":"","reg":false,"x":720,"y":360,"wires":[["9741c2a3.f204a"]]},{"id":"64b985f7.dac62c","type":"smooth","z":"59303127.37f18","name":"","property":"payload","action":"sd","count":"20","round":"0","mult":"single","reduce":false,"x":720,"y":420,"wires":[["9741c2a3.f204a"]]},{"id":"ce94b021.ab2a2","type":"change","z":"59303127.37f18","name":"","rules":[{"t":"set","p":"payload","pt":"msg","to":"register102","tot":"global"}],"action":"","property":"","from":"","to":"","reg":false,"x":560,"y":420,"wires":[["64b985f7.dac62c"]]},{"id":"78a1ae1.bdaf35","type":"change","z":"59303127.37f18","name":"","rules":[{"t":"set","p":"payload","pt":"msg","to":"register103","tot":"global"}],"action":"","property":"","from":"","to":"","reg":false,"x":580,"y":480,"wires":[["9741c2a3.f204a"]]},{"id":"3bc2e903.9b6b86","type":"random","z":"59303127.37f18","name":"","low":"0","high":"1000","inte":"true","property":"payload","x":420,"y":180,"wires":[["cbe29c71.d2dfa"]]},{"id":"cbe29c71.d2dfa","type":"smooth","z":"59303127.37f18","name":"","property":"payload","action":"mean","count":"10","round":"0","mult":"single","reduce":false,"x":560,"y":180,"wires":[["1aaa5da2.d79182"]]},{"id":"94b02819.8ff308","type":"modbus server","z":"","host":"localhost","port":"502","debug":false,"unitID":"1"},{"id":"ecc1bc26.f60bd","type":"modbustcp-server","z":"","name":"","host":"127.0.0.1","port":"502","unit_id":"1","reconnecttimeout":"3"}]`

- In regards to the leftover comment, it can be deleted. The error in this case works best as a string only so it is formatted correctly in the console/log.

Thank you for considering my code.

Reply:
Ok. If you are up to the challenge to implement the test cases using the mocha framework, I'd be happy about it. Otherwise I'll use the flows that you provided and implement a test myself before I merge.

See [test/modbus-in.spec.js](../blob/master/test/modbus-in.spec.js) for reference.